src/layouts/ScheduleContainer.jsx

```diff
@@ -5,18 +5,68 @@ import Schedule from "components/Schedule";
 import { useState } from "react";
 import { styled } from "@mui/material";
 import { ShareButton } from "components/ShareButton";
+import { useFormContext } from "context/Form";
+
+const rootURL = `${window.location.protocol}//${window.location.hostname}${
+  window.location.port ? ":" + window.location.port : ""
+}`;
+
+export const generateShareLink = ({
+  schedule,
+  scheduleTerm,
+  evening,
+  online,
+  startPref,
+  consecPref,
+  resultSize,
+  blacklist,
+}) => {
+  console.log(schedule);
+  const course_ids = schedule
+    .map((item) => `${item.objects.course}`)
+    .filter((course, index, self) => self.indexOf(course) === index)
+    .join(",");
+
+  const classes = schedule.map((item) => `${item.objects.class}`);
+  const blacklist_ids = Object.keys(blacklist)
+    .filter((id) => !classes.includes(id))
+    .join(",");
+
+  const eveningClassesBit = evening === true ? "1" : "0";
+  const onlineClassesBit = online === true ? "1" : "0";
+
+  return encodeURI(
+    rootURL +
+      "/" +
+      `?term=${scheduleTerm}&courses=[${course_ids}]&evening=${eveningClassesBit}&online=${onlineClassesBit}&start=${startPref}&consec=${consecPref}&limit=${resultSize}&blacklist=[${blacklist_ids}]`
+  );
+};
 
 const UnstyledScheduleContainer = ({
   className,
   courseOrder,
   schedules,
   aliases,
   errmsg,
-  shareLink,
 }) => {
+  const { values } = useFormContext();
   const [showInstructorNames, setShowInstructorNames] = useState(true);
   const [page, setPage] = useState(0);
 
+  let shareLink = "";
+  if (schedules.length > 0) {
+     shareLink = generateShareLink({
+      schedule: schedules[page],
+      scheduleTerm: values.scheduleTerm,
+      evening: values.evening,
+      online: values.online,
+      startPref: values.startPref,
+      consecPref: values.consecPref,
+      resultSize: values.resultSize,
+      blacklist: values.blacklist,
+    });
+  }
```

Comment:
This is outside any handlers and will re-run on any state change at this level or above, can this logic be moved to the `copyShareLink` handler? That way it will only run on shareButton click.

Reply:
I think that handler is within the formContext so you will have access to those values

Reply:
> I think that handler is within the formContext so you will have access to those values

@mdziubin This is fixed now, I have added it to copyShareLink, as well as fixed behaviour that prevented it from savig correctly.

My only concern is that doing it this way causes a slight delay in copying the schedule, whereas the other method allowed it to instantly copy.

Let me know if there are any other issues, I had to make it fetch each class to get all course data, so that I could blacklist every option that was not in the current image. I am not sure if there was any better way to do this though.

Reply:
If there is any delay in copy due to the `generateShareLink` that same delay would hit the rerender with the previous setup. You would feel it on each input change so this is the better option. Looking at the code for `generateShareLink` I don't see any reason it would have a noticable delay.

Reply:
The delay comes from the fetchClasses that have to wait on 1-6 api requests depending on how many classes are selected. Not super noticeable though. My thinking was that this small delay is just more noticeable when clicking Copy, than when submitting schedules and a delay is already expected. I think you're right though this is probably not noticeable enough to be a problem.

As for the async, I did it so I could await fetchClasses. Originally I tried something like this, however blacklist_ids was empty.

`  let allClasses = [];
  for (let i = 0; i < schedule.length; i++) {
    fetchClasses(scheduleTerm, schedule[i].objects.course).then((fetchedClasses) => {
      if (fetchedClasses.LEC !== undefined) {
        allClasses.push(...fetchedClasses.LEC.map((lec) => lec.id));
      }
      if (fetchedClasses.LAB !== undefined) {
        allClasses.push(...fetchedClasses.LAB.map((lab) => lab.id));
      }
      if (fetchedClasses.SEM !== undefined) {
        allClasses.push(...fetchedClasses.SEM.map((lab) => lab.id));
      }
    });
  }

  // Blacklist is all section IDs minus the classes in the schedule
  const blacklist_ids = allClasses.filter((id) => !classes.includes(id)).join(",");`